source/rtp_pkt_queue.c

```diff
@@ -0,0 +1,265 @@
+    }
+
+    if( result == RTP_PACKET_QUEUE_RESULT_OK )
+    {
+        if( IS_QUEUE_FULL( pQueue ) )
+        {
+            result = RTP_PACKET_QUEUE_RESULT_FULL;
+        }
+    }
+
+    if( result == RTP_PACKET_QUEUE_RESULT_OK )
+    {
+        pQueue->pRtpPacketInfoArray[ pQueue->writeIndex ].seqNum = pRtpPacketInfo->seqNum;
+        pQueue->pRtpPacketInfoArray[ pQueue->writeIndex ].pSerializedRtpPacket = pRtpPacketInfo->pSerializedRtpPacket;
+        pQueue->pRtpPacketInfoArray[ pQueue->writeIndex ].serializedPacketLength = pRtpPacketInfo->serializedPacketLength;
+
+        pQueue->writeIndex = INC_WRITE_INDEX( pQueue );
+        pQueue->packetCount += 1;
+    }
+
+    return result;
+}
+
+/*----------------------------------------------------------------------------*/
+
+/**
+ * @brief Add an RTP packet info into the specified queue.
+ *
+ * Delete the oldest RTP packet info and return it, if the queue is full.
+ */
+RtpPacketQueueResult_t RtpPacketQueue_ForceEnqueue( RtpPacketQueue_t * pQueue,
+                                                    const RtpPacketInfo_t * pRtpPacketInfo,
+                                                    RtpPacketInfo_t * pDeletedRtpPacketInfo )
+{
+    RtpPacketQueueResult_t result = RTP_PACKET_QUEUE_RESULT_OK;
+
+    if( ( pQueue == NULL ) ||
+        ( pRtpPacketInfo == NULL ) ||
+        ( pDeletedRtpPacketInfo == NULL ) )
+    {
+        result = RTP_PACKET_QUEUE_RESULT_BAD_PARAM;
+    }
+
+    if( result == RTP_PACKET_QUEUE_RESULT_OK )
+    {
+        if( IS_QUEUE_FULL( pQueue ) )
+        {
+            pDeletedRtpPacketInfo->seqNum = pQueue->pRtpPacketInfoArray[ pQueue->readIndex ].seqNum;
+            pDeletedRtpPacketInfo->pSerializedRtpPacket = pQueue->pRtpPacketInfoArray[ pQueue->readIndex ].pSerializedRtpPacket;
+            pDeletedRtpPacketInfo->serializedPacketLength = pQueue->pRtpPacketInfoArray[ pQueue->readIndex ].serializedPacketLength;
+
+            pQueue->readIndex = INC_READ_INDEX( pQueue );
+            pQueue->packetCount -= 1;
+            result = RTP_PACKET_QUEUE_RESULT_PACKET_DELETED;
+        }
+
+        pQueue->pRtpPacketInfoArray[ pQueue->writeIndex ].seqNum = pRtpPacketInfo->seqNum;
+        pQueue->pRtpPacketInfoArray[ pQueue->writeIndex ].pSerializedRtpPacket = pRtpPacketInfo->pSerializedRtpPacket;
+        pQueue->pRtpPacketInfoArray[ pQueue->writeIndex ].serializedPacketLength = pRtpPacketInfo->serializedPacketLength;
+
+        pQueue->writeIndex = INC_WRITE_INDEX( pQueue );
+        pQueue->packetCount += 1;
+    }
+
+    return result;
+}
+
+/*----------------------------------------------------------------------------*/
+
+/**
+ * @brief Read and remove the oldest RTP packet info from the queue.
+ */
+RtpPacketQueueResult_t RtpPacketQueue_Dequeue( RtpPacketQueue_t * pQueue,
+                                               RtpPacketInfo_t * pRtpPacketInfo )
+{
+    RtpPacketQueueResult_t result = RTP_PACKET_QUEUE_RESULT_OK;
```

Comment:
Is it possible to merge common part in `RtpPacketQueue_Dequeue()` and `RtpPacketQueue_Peek()` as they're pretty similar?

Reply:
RtpPacketQueue_Dequeue is removing the element from the queue , where as RtpPacketQueue_Peek is just returning the element. As the functionalities are different hence we are keeping both. 
But at later point of time while implementing application side code it is not required we can remove it then.

Reply:
Hi @moninom1,
I didn't mean to merge two API into one. What I mean is to merge the "common part" inside. What do you think?
For example:
```
RtpPacketQueue_Peek()
{
    /* Peek the element. */
}

RtpPacketQueue_Dequeue()
{
    RtpPacketQueue_Peek();

    /* removing that element from the queue. */
}
```

Reply:
Yes I agree with @ActoryOu .  I see also right now there is not consistent behavior in these two APIs if the packet info struct is NULL.  IN peek we will return bad param but in dequeue we will still update the readIndex and packetCount on the queue even when the packet info is NULL.  Do we intend for this divergence?

Reply:
Sure let me update that.